From 92b009babd1e4a63dd34638924b9559727713369 Mon Sep 17 00:00:00 2001
From: Leah Rowe <leah@osboot.org>
Date: Mon, 22 Feb 2021 22:16:59 +0000
Subject: [PATCH 04/65] lenovo/t60: make 64MiB VRAM the default in cmos.default

---
 src/mainboard/lenovo/t60/cmos.default | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/mainboard/lenovo/t60/cmos.default b/src/mainboard/lenovo/t60/cmos.default
index 283a5747ee..91f6c0e6e2 100644
--- a/src/mainboard/lenovo/t60/cmos.default
+++ b/src/mainboard/lenovo/t60/cmos.default
@@ -17,4 +17,4 @@ trackpoint=Enable
 sticky_fn=Disable
 power_management_beeps=Enable
 low_battery_beep=Enable
-gfx_uma_size=8M
+gfx_uma_size=64M
-- 
2.39.5

